Level 1

beagle-elgaeb opened the pull request:
Бэкенд дипломного проекта

Commit:
ДП-1. Выполнены подготовительные действия:
- клонирован репозиторий;
- создана ветка для первого этапа;
- добавлены необходимые инфраструктурные файлы;
- инициализирован package.json.

Commit:
ДП-1. Созданы схемы и модели ресурсов API:
- user;
- movie.

Commit:
ДП-1. Созданы роуты и контроллеры:
- GET /users/me;
- PATCH /users/me;
- GET /movies;
- POST /movies;
- DELETE /movies/:movieId.

Commit:
ДП-1. Реализовано логирование.

Commit:
ДП-1. Реализованы аутентификация и авторизация:
- POST /signup;
- POST /signin;
- POST /signout.
Роуты проверены постманом.

Commit:
ДП-1. Рефакторинг и переписано на TS.

Commit:
ДП-1. Отредактирован REDAME.

Commit:
ДП-1. Мелкие правки.

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
    - [x] в проекте есть `package.json`;
    - [x] `.editorconfig`;
    - [x] `.eslintrc`, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
_- Лишние неиспользуемые пакеты следует удалить из package.json_
![image](https://user-images.githubusercontent.com/17438227/142917947-8b3c99da-3343-4d8c-b946-06596addbf0e.png)
    - [x] `.gitignore`;
    - [x] в `.eslintrc` добавлено исключение для `_id`. Директивы `eslint-disable`, `eslint-disable-line` и `eslint-disable-next-line` запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды `npx eslint .` они должны отсутствовать. 
- [x] <!--3.64--> В разделе scripts файла `package.json` есть команды:
    - [x] `npm run start`, которая запускает сервер на `localhost:3000`;
    - [x] `npm run dev`, которая запускает сервер на `localhost:3000` с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно: 
    - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
    - [x] `PATCH /users/me` — обновляет информацию о пользователе;
_- По заданию маршрут PATCH /users/me должен обновлять почтовый ящик и имя пользователя._
    - [x] `GET /movies` — все сохранённые пользователем фильмы;
    - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
_- Создание фильма требует передачи свойства movieId._
_- Поле movieId должно быть числом в десятичном формате. api https://api.nomoreparties.co/beatfilm-movies возвращают фильмы, у которых id (moviId для фильмов в нашей базе данных) является числовым значением. Пожалуйста, не забудьте поменять тип поля в схеме модели и в схеме валидации celebrate._
    - [x] `DELETE /movies/:id` — удаляет сохранённый фильм по `_id`;
    - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
    - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
    - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя. 
- [x] <!--3.64--> Все роуты, кроме `/signin` и `/signup`, защищены авторизацией.
_- Маршрут по умолчанию не защищён авторизацией._
_В случае обращения к защищенному авторизацией маршруту незалогиненным пользователем должна возвращаться ошибка со статусом 401._
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются: 
    - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
    - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
_- При обновлении данных пользователя с использованием почтового ящика, который принадлежит другому юзеру, необходимо возвращать ошибку 409._
_Сейчас возвращается ошибка 500._
    - [x] сообщение об ошибке соответствует ее типу;
    - [x] асинхронные обработчики завершены блоком `catch`;
    - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
    - [x] пароль хранится в виде хеша;
    - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
    - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
_- Попытка создать пользователя без имени не прерывается celebrate и уходит контроллеру. В схеме celebrate для всех полей необходимо указывать обязательное их наличие в запросе._
    - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из `process.env`.
- [x] <!--3.64--> К серверу можно обратиться с помощью https по адресу домена, указанному в README.md.
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT: 
    - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
    - [x] в режиме разработки (при  `process.env.NODE_ENV !== 'production'`) код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле `index.js`, который находится в папке `routes`. Оттуда единый роут подключается в файле `app.js`. 
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
    - [x] запросы и ответы записываются в файл `request.log`;
    - [x] ошибки записываются в файл `error.log;`
_- Ответ с ошибкой сразу отправляется на клиент без генерации объекта ошибки, которая должна передаваться по цепочке обработки запроса. Эта ошибка не будет добавлена в лог ошибок._
    - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone.
    Подходящий адрес: movies-explorer.nomoreparties.co/api.
    Неподходящий адрес: movies-explorer.nomoreparties.co.

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор `Error`. 
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью. 
- [x] <!--0.71--> Конфигурация и константы в отдельных файлах: 
    - [x] адрес Mongo-сервера, секретный ключ для JWT в режиме разработки, вынесены в отдельный конфигурационный файл;
    - [x] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено. 
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co. 
- [x] <!--0.71--> Для имени базы данных используется `moviesdb` 

## Количество баллов: 47

## Доп. Комментарии:
- Все мидлвары, кроме requestLogger, стоит подключить после limiter. Если limiter отклонит запрос от обработки, то эти мидлвары, подключенные ранее, выполнят лишнюю работу.

Commit:
ДП-1. Исправлено по замечаниям ревьюера:
- неиспользуемые пакеты удалены из package.json;
- маршрут PATCH /users/me обновляет почтовый ящик;
- при создании фильма передаётся свойство movieId;
- маршрут по умолчанию защищён авторизацией;
- при обновлении данных пользователя с использованием почтового ящика, который принадлежит другому юзеру, возвращается ошибка 409;
- попытка создать пользователя без имени прерывается celebrate и не уходит контроллеру;
- мидлвары, кроме requestLogger, подключены после limiter.

Commit:
ДП-1. Перешла в локальной версии на npm

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
    - [x] в проекте есть `package.json`;
    - [x] `.editorconfig`;
    - [x] `.eslintrc`, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
    - [x] `.gitignore`;
    - [x] в `.eslintrc` добавлено исключение для `_id`. Директивы `eslint-disable`, `eslint-disable-line` и `eslint-disable-next-line` запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды `npx eslint .` они должны отсутствовать. 
- [x] <!--3.64--> В разделе scripts файла `package.json` есть команды:
    - [x] `npm run start`, которая запускает сервер на `localhost:3000`;
    - [x] `npm run dev`, которая запускает сервер на `localhost:3000` с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно: 
    - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
    - [x] `PATCH /users/me` — обновляет информацию о пользователе;
    - [x] `GET /movies` — все сохранённые пользователем фильмы;
    - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
    - [x] `DELETE /movies/:id` — удаляет сохранённый фильм по `_id`;
    - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
    - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
    - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя. 
- [x] <!--3.64--> Все роуты, кроме `/signin` и `/signup`, защищены авторизацией.
_- Маршрут по умолчанию не защищён авторизацией._
_В случае обращения к защищенному авторизацией маршруту незалогиненным пользователем должна возвращаться ошибка со статусом 401._
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются: 
    - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
    - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
    - [x] сообщение об ошибке соответствует ее типу;
    - [x] асинхронные обработчики завершены блоком `catch`;
    - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
    - [x] пароль хранится в виде хеша;
    - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
    - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
_- Попытка создать пользователя без имени не прерывается celebrate и уходит контроллеру. В схеме celebrate для всех полей необходимо указывать обязательное их наличие в запросе._
    - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из `process.env`.
- [x] <!--3.64--> К серверу можно обратиться с помощью https по адресу домена, указанному в README.md.
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT: 
    - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
    - [x] в режиме разработки (при  `process.env.NODE_ENV !== 'production'`) код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле `index.js`, который находится в папке `routes`. Оттуда единый роут подключается в файле `app.js`. 
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
    - [x] запросы и ответы записываются в файл `request.log`;
    - [x] ошибки записываются в файл `error.log;`
    - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone.
    Подходящий адрес: movies-explorer.nomoreparties.co/api.
    Неподходящий адрес: movies-explorer.nomoreparties.co.

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор `Error`. 
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью. 
- [x] <!--0.71--> Конфигурация и константы в отдельных файлах: 
    - [x] адрес Mongo-сервера, секретный ключ для JWT в режиме разработки, вынесены в отдельный конфигурационный файл;
    - [x] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено. 
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co. 
- [x] <!--0.71--> Для имени базы данных используется `moviesdb` 

## Количество баллов: 100

Commit:
ДП-1. Добавлен REDAME.
